azurelinuxagent/ga/cgroupapi.py

```diff
@@ -103,28 +109,99 @@ def _foreach_legacy_cgroup(operation):
+        If the result is 'tmpfs', cgroup v1 or a hybrid mode is being used.
+            If the result of 'stat -f --format=%T /sys/fs/cgroup/unified' is 'cgroup2fs', then hybrid mode is being used.
+
+    Raises exception if cgroup filesystem mountpoint is not '/sys/fs/cgroup', or an unknown mode is detected. Also
+    raises exception if hybrid mode is detected and there are controllers available to be enabled in the unified
+    hierarchy (the agent does not support cgroups if there are controllers simultaneously attached to v1 and v2
+    hierarchies).
+    """
+    if not os.path.exists(CGROUP_FILE_SYSTEM_ROOT):
+        v1_mount_point = shellutil.run_command(['findmnt', '-t', 'cgroup', '--noheadings'])
+        v2_mount_point = shellutil.run_command(['findmnt', '-t', 'cgroup2', '--noheadings'])
+        raise InvalidCgroupMountpointException("Expected cgroup filesystem to be mounted at '{0}', but it is not.\n v1 mount point: \n{1}\n v2 mount point: \n{2}".format(CGROUP_FILE_SYSTEM_ROOT, v1_mount_point, v2_mount_point))
+
+    root_hierarchy_mode = shellutil.run_command(["stat", "-f", "--format=%T", CGROUP_FILE_SYSTEM_ROOT]).rstrip()
+
+    if root_hierarchy_mode == "cgroup2fs":
+        log_cgroup_info("Using cgroup v2 for resource enforcement and monitoring")
+        return SystemdCgroupApiv2()
+
+    elif root_hierarchy_mode == "tmpfs":
+        # Check if a hybrid mode is being used
+        unified_hierarchy_path = os.path.join(CGROUP_FILE_SYSTEM_ROOT, "unified")
+        if os.path.exists(unified_hierarchy_path) and shellutil.run_command(["stat", "-f", "--format=%T", unified_hierarchy_path]).rstrip() == "cgroup2fs":
+            # Hybrid mode is being used. Check if any controllers are available to be enabled in the unified hierarchy.
+            available_unified_controllers_file = os.path.join(unified_hierarchy_path, "cgroup.controllers")
+            if os.path.exists(available_unified_controllers_file):
+                available_unified_controllers = fileutil.read_file(available_unified_controllers_file).rstrip()
+                if available_unified_controllers != "":
+                    raise CGroupsException("Detected hybrid cgroup mode, but there are controllers available to be enabled in unified hierarchy: {0}".format(available_unified_controllers))
+
+        cgroup_api = SystemdCgroupApiv1()
+        # Previously the agent supported users mounting cgroup v1 controllers in locations other than the systemd
+        # default ('/sys/fs/cgroup'). The agent no longer supports this scenario. If either the cpu or memory
+        # controller is mounted in a location other than the systemd default, raise Exception.
+        if not cgroup_api.are_mountpoints_systemd_created():
+            raise InvalidCgroupMountpointException("Expected cgroup controllers to be mounted at '{0}', but at least one is not. v1 mount points: \n{1}".format(CGROUP_FILE_SYSTEM_ROOT, json.dumps(cgroup_api.get_controller_root_paths())))
```

Comment:
we want to do the cleanup only for v1, but InvalidCgroupMountpointException may also be raised for other reasons and the caller has no way to know why it was raised. maybe use a different exception? InvalidV1CgroupMountpointException

Reply:
InvalidCgroupMountpointException is only raised if v1 controllers are not systemd default or if /sys/fs/cgroup path does not exist 

I think we want to do cleanup in both cases. Previously we supported users not using /sys/fs/cgroup. So if the path doesn't exist, we should do agent drop in cleanup in case we were supporting cgroups at a different path before

Reply:
`or if /sys/fs/cgroup path does not exist
`
this could happen under v2 and there is no need to cleanup in that case

Reply:
That's true but I think it is safer to also cleanup those files on v2 machines which meet this condition than to keep the files on v1 machines which meet this condition.

Reply:
do you mean we have been creating those files under v2 as well? i thought this cleanup was v1-specific

Reply:
yes drop in files are currently being created on v2 machines, with the exception of the CPUQuota drop-in file. This is because in cgroupconfigurator.initialize the azure slice/drop-in files are created (__setup_azure_slice) before the controller mountpoints are checked. The CPUQuota drop in file is only created after the controller mountpoints are checked. As a result, v2 machines have cpu/memory accounting drop-in files. 

Since quota drop-in file is not created on v2 machines, then it is not necessary to cleanup drop-in files on v2 (in my opinion). But, if /sys/fs/cgroup path does not exist, then we don't know whether the machine is using v1 or v2, so we should clean up the drop-in files regardless. If the machine is using v1, we should definitely clean up the drop in files since quota could exist. If the machine is using v2, there is no harm in cleaning up the drop- in files.

Reply:
BTW with this PR, the azure slice/drop in files are not created until after we check if v2 is in use. So on new machines with this update, drop-in files won't be created if v2 is in use.

Reply:
ok, sounds good then, thanks